### DISCUSSION TITLE
React-flow for Angular

### DISCUSSION THREAD

**Discussion by weilies** [xyflow/xyflow#2012] [General]
Dear support,

I am from Angular with no react background. Looking to expand a web app (angular) with react-flow.
Since it's typescript built, can anyone share the sample code if they able to run this in their angular app rather than react app?

Thanks in million!

**Comment by jmls** [comment]
did you ever get to implement this ? I have an angular 15 app that I would love to use this component in

**Reply by knackstedt** [reply]
Check out #4887 :)

**Comment by Andrewknackstedt** [comment]
I ended up writing a wrapper for Angular on my test app -- it works but you need to configure your tsconfig and install react.
https://github.com/knackstedt/angular-experiments/tree/main/src/app/pages/%40react/reactflow

**Reply by moklick** [reply]
🔥🔥🔥

**Reply by boradakash** [reply]
Oh, that's great, thanks!
@Andrewknackstedt is there any limitation you find using it as a wrapper?

**Reply by knackstedt** [reply]
I'm still working on some of the kinks, it doesn't immediately support angular components as reactflow nodes.

This is a side-project for me, so I haven't been dedicating lots of time to it

**Reply by boradakash** [reply]
Cool thanks!

**Reply by knackstedt** [reply]
Here is a cleaned up sample:
Source: https://github.com/knackstedt/dotglitch.dev/tree/main/src/app/pages/%40showcase/reactflow-demo
Demo: https://dotglitch.dev/#/Reactflow

**Reply by knackstedt** [reply]
#4887

**Comment by knackstedt** [comment]
Please check out #4887 for my updated release with **much** simpler documentation

**Comment by mukulhy** [comment]
There is a jsplumb library in angular that is almost same as react flow

**Reply by aloksharma1** [reply]
jsplumb toolkit is paid, and the ui looks archaic do you have any other suggestions?

**Reply by mukulhy** [reply]
I used in a app its has 2 version paid and a free one i used free one apart from that right now i dont have alternate if you find a better one please share here as well

**Reply by knackstedt** [reply]
https://github.com/xyflow/xyflow/discussions/4887

**Comment by shyamal890** [comment]
Would love to see an Angular library from xyflow team. Given the new signals and standalone components it should feel very similar to what react has to offer.

**Reply by markwojno** [reply]
https://github.com/Foblex/f-flow

**Reply by aloksharma1** [reply]
> https://github.com/Foblex/f-flow

good find

**Reply by knackstedt** [reply]
https://github.com/xyflow/xyflow/discussions/4887

**Comment by ritesh-afferent** [comment]
I have used [Rete JS](https://retejs.org/) on Angular and it's not bad. But I agree with shyamal890, it would be really great to have an Angular library supported by the xyflow team.

**Comment by stephengardner** [comment]
I've tried everything now.  jsplumb angular integration just straight up doesn't render, and when it does, the drag/drop functionality breaks it immediately, it's very wonky.

the ngx-reactflow didn't seem to obey dragging/dropping either.

foblex is completely counter-intuitive and doesn't follow any reasonable nodes/edges paradigm.

not much left to try.

really hope xyflow gets something going for angular

**Reply by knackstedt** [reply]
Can you reference a specific drag and drop example you were looking at? I can check if it's possible to add support in my updated version

**Comment by knackstedt** [comment]
I'm coming back with a much more robust implementation. Still in a beta state where some CD events are missed but I'm open to PRs to resolve those as well as improve support going forward.
I updated it to use their new xyflow naming system, and rebuilt the implementation so that everything can be addressed through a template. A big reason I did this is because it makes custom nodes **much** less of a hassle to write and maintain -- no more hacks needed for basic functionality.
Package: https://www.npmjs.com/package/ngx-xyflow
Source: https://github.com/knackstedt/ngx-xyflow

**Reply by msahalkc** [reply]
Hi @knackstedt can you share a sample application that uses this package. Thanks in advance.

**Reply by knackstedt** [reply]
@msahalkc I don't have a large sample application yet -- I'm doing a huge overhaul on my main repo that uses this, though the package has a working basic example. 
https://github.com/knackstedt/ngx-xyflow/tree/main/src/app

**Reply by BenjaminHofstetter** [reply]
I didn't check it but this looks similar
https://www.ngdiagram.dev

**Comment by siarheihuzarevich** [comment]
Foblex Flow maintainer here. Since this thread ranks highly for “React Flow for Angular”, here is the current API and a transparent comparison for anyone landing here.

Foblex uses Angular templates and directives rather than requiring the editor UI to mirror a `nodes[]` / `edges[]` configuration model. Nodes remain ordinary Angular components with application-owned bindings. A minimal editor with an existing connection and interactive connection creation looks like this:

```html
<f-flow fDraggable (fCreateConnection)="onConnect($event)">
  <f-connection-for-create />

  <f-canvas>
    <f-connection fSourceId="a" fTargetId="b" />

    <div fNode fDragHandle [fNodePosition]="p1">
      A
      <div fConnector fConnectorId="a" fConnectorType="source"></div>
    </div>

    <div fNode fDragHandle [fNodePosition]="p2">
      B
      <div fConnector fConnectorId="b" fConnectorType="target"></div>
    </div>
  </f-canvas>
</f-flow>
```

Foblex covers the core editor surface most Angular teams need: selection, zoom/pan, minimap, Dagre/ELK layout adapters, progressive virtualization, and optional managed state. React Flow itself has built-in accessibility support; Foblex v19 likewise provides ARIA semantics and keyboard navigation, and also supports keyboard-driven connection creation.

The choice is mainly architectural: Foblex is template/directive-first and Angular-native; ngx-vflow is another legitimate Angular option if a nodes/edges array model feels more natural.

Current comparison, including tradeoffs: https://flow.foblex.com/docs/react-flow-vs-foblex-flow-for-angular-teams
